### PR TITLE
added a divide by 0 failsafe

### PR DESCRIPTION
**Pull request by TitusBraber**
Referencing back to https://github.com/paparazzi/paparazzi/issues/2097#issuecomment-320302913 again, there is still a divide by 0 possible and actually happening.
If for all samples in  
for (uint16_t sample = 0; sample < n_samples; sample++)
 distance_1 is 0, then the newly introduced used_samples remains 0, leading to a division by 0 in the mean_f function in:
https://github.com/paparazzi/paparazzi/blob/master/sw/airborne/modules/computer_vision/opticflow/size_divergence.c#L118
This value then messes up filters where the size_divergence is used, for example OF_Landing and forthcoming OF_Hover.

The suggested fix sets the mean_divergence to 0. Though I am NOT sure if this is the correct behavior to fix it.

### CODE REVIEW

**Comment by kirkscheper** [comment]
You not put the failsafe on the mean function?

**Comment by TitusBraber** [comment]
I thought about that @kirkscheper, I even looked if the function was being used by other files (it is not).
I wasn't sure, but judging from that it doesn't cause a segfault, instead returning a NaN, it is a float division, I thought it might be better to have this for people that might use that function and not expect a 0 entries gives 0 result. Also I wasn't sure if an if check in a math function was preferred.

If the putting the failsafe on the mean function is the preferred behavior I will change it to that instead, and include a comment above the mean_f function to describe the behavior. What do you think?

**Comment by kirkscheper** [comment]
When is nan ever good on a flying vehicle? Can you update this?

**Comment by kirkscheper** [comment]
If your drone is falling out of the sky due to a Nan I don't think that it is very useful for debugging. I don't really care what you do to show debugging info to the user, up to you.

File: sw/airborne/math/pprz_stat.c [hunk 1]
```diff
@@ -73,18 +78,22 @@ int32_t variance_i(int32_t *array, uint32_t n_elements)
...
-  // Determine the covariance:
-  uint32_t i;
-  for (i = 0; i < n_elements; i++) {
-    sumX += (float)array1[i];
-    sumY += (float)array2[i];
-    sumXY += (float)(array1[i]) * (float)(array2[i]);
+  if (n_elements == 0) {
+    // Note that something else is wrong if you want the covariance of 0 samples.
+    return 0;
+  } else {
```
Comment (kirkscheper) [inline sw/airborne/math/pprz_stat.c]: Writing it like this is not good practice as the second return only happens if the else is called but the return uses things that are defined inside the else, not good.

You should have the if with the n_elements then the rest follow without the else and brackets.

File: sw/airborne/modules/computer_vision/opticflow/size_divergence.c [hunk 2]
```diff
@@ -114,8 +114,17 @@ float get_size_divergence(struct flow_t *vectors, int count, int n_samples)
...
   }
 
-  // calculate the mean divergence:
-  float mean_divergence = mean_f(divs, used_samples);
+  float mean_divergence;
+  // calculate the mean divergence or use 0 when used_samples is still 0
+  if(used_samples==0)
+  {
+    // Note that this shouldn't happen
+    mean_divergence = 0.0;
```
Comment (kirkscheper) [inline sw/airborne/modules/computer_vision/opticflow/size_divergence.c]: is this necessary?

File: sw/airborne/math/pprz_stat.c [hunk 3]
```diff
@@ -140,16 +154,20 @@ float variance_f(float *array, uint32_t n_elements)
...
-
-  // Determine the covariance:
-  uint32_t i;
-  for (i = 0; i < n_elements; i++) {
-    sumX += arr1[i];
-    sumY += arr2[i];
-    sumXY += arr1[i] * arr2[i];
+  if (n_elements == 0) {
+    // Note that something else is wrong if you want the covariance of 0 samples.
+    return 0.0;
```
Comment (kirkscheper) [inline sw/airborne/math/pprz_stat.c]: this needs to return a float, 0.0 is a double, use 0.f

**Comment by TitusBraber** [comment]
Thanks @kirkscheper for the feedback and explanation. I implemented all proposed fixes.

File: sw/airborne/modules/computer_vision/opticflow/size_divergence.c [hunk 4]
```diff
@@ -114,8 +114,9 @@ float get_size_divergence(struct flow_t *vectors, int count, int n_samples)
     }
   }
 
-  // calculate the mean divergence:
-  float mean_divergence = mean_f(divs, used_samples);
+  float mean_divergence;
+  // calculate the mean divergence or use 0 when used_samples is still 0
+  mean_divergence = mean_f(divs, used_samples);
```
Comment (kirkscheper) [inline sw/airborne/modules/computer_vision/opticflow/size_divergence.c]: why this change?

Comment (TitusBraber) [reply sw/airborne/modules/computer_vision/opticflow/size_divergence.c]: It looks like that is the remainder from when I fixed it in size_div. Instead of unfixing it I changed it again. Next commit should restore it to the original state.
